### ISSUE TITLE
Write Tour of the Project, including OSH compatibility and Oil clean slate

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#1127]
Somehow after years of writing it's not clear to people that OSH is both POSIX compatible and the most bash compatible shell by a mile

https://lobste.rs/s/jdxhu9/sqlite_go_with_without_cgo#c_fka0da

**Comment by bar-g** [comment]
Here is some suggestion (draft):

Oilshell is an improved Unix shell.  (An upgrade path from {a,ba,da,k,z}sh to a better language and runtime, even amicable for Python and JavaScript users who avoid shell.)

By default, it is "pretty"-compatible with other shells (/bin/osh).
* It runs most existing scripts without having to enable any ugly backwards-compatibility option.
* It allows enabling individual strict_* modes (or strict:all) to ensure robustness of shell scripts (may require script adjustments).
* It allows adding "drops" of oil syntax to existing scripts (oil:allow?), enabling nice solutions for legacy shell problems but won't work without some special support in other shells.
* Enabling the full language (oil:full?) will only disable support for some conflicting or unclear?, superseded legacy shell syntax.

However, new scripts may also be written straightly with all the improved shell concepts of the Oil language enabled, and all superseded shell syntax disabled (/bin/oil and oil:only?).

**Comment by bar-g** [comment]
Initial confusion might also be reducable by more carefully, and explicitly distinguishing between oil/oil-language/oilshell/osh.
Avoiding to talk about just "oil" instead of a more precise meaning: "oil language", or "oilshell interpreter", or meaning "osh mode", etc.

project and interpreter (oilshell?)
mode to support "other shell"s scripts (osh?)
new language and mode (oil?)

**Comment by bar-g** [comment]
* Add link to https://github.com/oilshell/oil/wiki/Language-Design-Principles on homepage

**Comment by bar-g** [comment]
I updated the intro/mini-tour for the homepage in the above comment https://github.com/oilshell/oil/issues/1127#issuecomment-1133749799

 - [ ] Also add missing direct link to the documentation on the homepage.

**Comment by andychu** [comment]
FWIW I put a preview here of something with MANY FEWER words:

https://www.oilshell.org/index2.html

again I think  you're overestimating how much people will read

----

But I am not entirely happy with this either.  

I agree there is some "Oil" confusion, but I think "Oilshell" is not distinct enough.

Even if we add "Oilshell", "Oil" is still going to refer to either the language or the project.

Similar to "Go" and "Golang" which people use to refer to both the language and project (although there is only one language in that case.)

----

A different although unlikely solution is to rename the Oil language.  Oil the project could be both OSH and "Tea" or "Soil" (those are names I used for other subprojects, so they are not available, but give you an idea)

I think this would be a lot of effort and make all previous blog entries confusing ...  But possibly less confusing than introducing a similar-but-different "Oilshell" term

**Comment by andychu** [comment]
Existing tagline

> Oil is a new Unix shell. It's our upgrade path from [bash](https://www.oilshell.org/cross-ref.html?tag=bash#bash) to a better language and runtime. It's also for Python and JavaScript users who avoid shell!

----------------

That draft was

> A new Unix shell with 2 parts:
>
> 1.  OSH is the most bash-compatible shell by a mile!
> 1. Oil is our upgrade path from [bash](https://www.oilshell.org/cross-ref.html?tag=bash#bash) to a better language and runtime. It's also for Python and JavaScript users who avoid shell.

----------

Here is another one:


> Oil is a new Unix shell.  It runs your existing shell scripts, but it's also a new language for Python and JavaScript users who avoid shell.

> That is, it's our upgrade path from [bash](https://www.oilshell.org/cross-ref.html?tag=bash#bash) to a better language and runtime.

**Comment by andychu** [comment]
OK I updated the home page:

https://www.oilshell.org/

> Oil is a new Unix shell. It runs your existing shell scripts, and it's a new language for Python and JavaScript users who avoid shell!

> It's our upgrade path from [bash](https://www.oilshell.org/cross-ref.html?tag=bash#bash) to a better language and runtime.

I think this is good because we can have three links:

- "runs your existing shell scripts" -- link a  demonstration, e.g. along the lines of https://www.oilshell.org/blog/2018/01/15.html
- "new language" -- link to the tour
- "upgrade path" -- something like the table you made and https://github.com/oilshell/oil/wiki/Gradually-Upgrading-Shell-to-Oil

And now we have `shopt --set oil:ugprade`, which is very easy to remember!  (thanks for that feedback)

**Comment by bar-g** [comment]
Thanks, much appreciated. Let me add my thoughts, and some background reasoning.

As Oilshell is the domain name as well as the prospective package name I think it's rather an omission or oversight not to use it for in explaining. Not confusing at all, rather necessary to mention.

Shell users, especially early adopters, I'd expect to actually appreciate helpful condensed information, and not being even slightly afraid of extensive docs at all, but happy not having to go through blogs and docs just for the basics.

> {a,ba} stuff is way too clever and jargon-y

Yeah, let's drop that.

> "very" in quotes

same

> "scare" of using a new shell

That's actually a specific background reasoning for mentioning things like "unneeded backwards-compat options", "strict*", and "run under multiple shells" hints.

Rather than seeing these as a distraction see them as valuable bits of info, mentioned to assure readers it's taken care of, providing confidence to start downloading and playing with oilshell. Things for them to check off as covered, instead of leaving them wary of unforeseen compatibility problems etc., waiting to hit them. The "run under other shells" hints are especially related to assessing lock-in.

> I don't understand "dropwise"

I wanted to describe adding some new features to a script and using oil for the additions, but keeping the other parts mostly unchanged.

Concerning the length of the intro, I really don't think anyone looking into a new shell would be scared off by a postcard sized executive summary, rather the contrary. (Especially for pre 1.0 versions.) Though, surely there are still be improvements possible.

**Comment by bar-g** [comment]
>I think this is good because we can have three links:

Well, I'd say it scatters information instead of gathering and presenting a summary.

**Comment by andychu** [comment]
Yeah I think they can actually all link to three sections of the same project tour:

- run existing shell scripts
- new langauge
- upgrade path

**Comment by bar-g** [comment]
Note that when you don't explain things a bit and tell new users, for example, there are (minimal) things that will break in osh, you'll likely loose them before you know. Ah it's broken->bye.

**Comment by bar-g** [comment]
> But I am not entirely happy with this either.
> I agree there is some "Oil" confusion, but I think "Oilshell" is not distinct enough.

Then maybe you might have to look for a name, maybe something like Oilscript?

Regarding the intro, I think, yep, having the intro (only) on a single(!), but multi-linked, second summary-page could work, too.
(If you really prefer a no-text, nav-only homepage.)

**Comment by andychu** [comment]
The idea is that most scripts run under bash and OSH at the same time, so there isn't any real comparison between OSH->YSH and bash->YSH

I think most of the upgrade stuff has to be driven by users ... We definitely need better docs, describing the rough outline

But so far I don't see evidence that people want to sit down and read a bunch of docs and painstakingly upgrade their scripts.  So far the people interested in OSH are actually kinda distinct from the people who are interested in YSH

i.e. some people want only the old stuff, some people want only the new stuff

I'm sure there are people who want to upgrade, but I haven't heard from them, and there are many other things to do on the project

---

In 2025 the main focus will be YSH and Hay, and making them stable.    And then USING them for new projects - `#projects-with-oils`

Koichi has also done some great work on bash compatibility recently, so OSH is making good progress too

So the upgrade path exists, but it definitely takes manual work, and I expect that we'll see movement on it if people come on Zulip and say  "I have a 1000 line bash script" and tell us what problems they have
